in-progress/NNNN-off-chain-contract-registry.md

```diff
@@ -0,0 +1,44 @@
+|                      |                             |
+| -------------------- | --------------------------- |
+| Issue                | Off-chain contract registry |
+| Owners               | @spalladino                 |
+| Approvers            |                             |
+| Target Approval Date |                             |
+
+## Executive Summary
+
+Since private contracts are not broadcasted on-chain, they need to be registered in a PXE before they can be executed. We expect this to be typically done by the dapp in context, but for composability's sake and ease of development, especially during devnet, we could make things easier by keeping a canonical off-chain registry of contracts.
+
+## Introduction
+
+When executing a private function on a contract in the PXE, we require that the corresponding contract instance and artifact to have been registered in advance.
+
+We can extend the PXE to fall back to requesting the instance and artifact to an external provider, and registering them on the fly during a simulation. Since the contract address acts as a commitment to the instance and artifact, the PXE can validate the downloaded artifact against the address itself by recomputing the class id and instance address.
+
+## Interface
+
+We split the interface into two: one for uploading and registering instances and artifacts, and another for querying. This allows us to implement a trusted registry with just the query interface first, to make it quickly available to devs, and manage the registration centrally until we implement the more complex uploading API.
+
+### Querying
+
+#### GET `/contracts/{address}.json`
+
+Returns the JSON-serialized contract instance for the given address, which includes the class identifier.
```

Comment:
Remind me again what woulod go in here? Is it simply the class id, the initialisation args/hash, deployer address etc?

Reply:
Exactly! 

```ts
salt
deployer
contractClassId
initializationHash
publicKeysHash
```